AK/IpAddressCidr.h

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2024, famfo <famfo@famfo.xyz>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/IPv4Address.h>
+#include <AK/IPv6Address.h>
+#include <AK/Result.h>
+
+namespace AK {
+
+class [[gnu::packed]] IPv4AddressCidr;
+class [[gnu::packed]] IPv6AddressCidr;
+template<typename Address>
+class [[gnu::packed]] AddressTraits;
+
+template<OneOf<IPv4AddressCidr, IPv6AddressCidr> AddressFamily>
+class [[gnu::packed]] IPAddressCidr {
+public:
+    enum IPAddressCidrError {
+        CidrTooLong = 0,
+        StringParsingFailed,
+    };
+
+    static Result<AddressFamily, IPAddressCidrError> Create(AddressTraits<AddressFamily>::IpAddress address, u8 length)
+    {
+        if (length > AddressFamily::MAX_LENGTH)
+            return IPAddressCidrError::CidrTooLong;
+
+        return AddressFamily(address, length);
+    }
+
+    AddressTraits<AddressFamily>::IpAddress ip_address() const { return m_address; }
```

Comment:
I’m fairly sure you want to alias the dependent IP address type with `using IPAddress = AddressTraits<AddressFamily>::IPAddress` at the start of the public section of this base class.